dotCMS/src/main/java/com/dotcms/security/secret/ServiceIntegrationAPIImpl.java

```diff
@@ -0,0 +1,525 @@
+            throws DotDataException, DotSecurityException {
+        final Optional<ServiceSecrets> secretsForService = getSecrets(serviceKey, host,
+                user);
+        if (secretsForService.isPresent()) {
+            //The secret already exists and wee need to update one specific entry they rest should be copy as they are
+            final ServiceSecrets serviceSecrets = secretsForService.get();
+            final Map<String, Secret> secrets = serviceSecrets.getSecrets();
+            final ServiceSecrets.Builder builder = new ServiceSecrets.Builder();
+
+            for (final Entry<String, Secret> entry : secrets.entrySet()) {
+                //Replace all existing secret/property that must be copied as it is.
+                builder.withSecret(entry.getKey(), entry.getValue());
+            }
+            //finally override or add the new value.
+            builder.withSecret(keyAndSecret._1, keyAndSecret._2);
+
+            saveSecrets(builder.withServiceKey(serviceKey).build(), host, user);
+        } else {
+            //The secret is brand new . We need to create a whole new one.
+            final ServiceSecrets.Builder builder = new ServiceSecrets.Builder();
+            builder.withSecret(keyAndSecret._1, keyAndSecret._2);
+            saveSecrets(builder.withServiceKey(serviceKey).build(), host, user);
+        }
+    }
+
+    @Override
+    public void saveSecrets(final ServiceSecrets secrets, final Host host, final User user)
+            throws DotDataException, DotSecurityException {
+        if (userDoesNotHaveAccess(user)) {
+            throw new DotSecurityException(String.format(
+                    "Invalid secret update attempt on `%s` performed by user with id `%s` and host `%s` ",
+                    secrets.getServiceKey(), user.getUserId(), host.getIdentifier()));
+        } else {
+            final String internalKey = internalKey(secrets.getServiceKey(), host);
+            secretsStore.saveValue(internalKey, toJsonString(secrets).toCharArray());
+        }
+    }
+
+    @Override
+    public void deleteSecrets(final String serviceKey, final Host host, final User user)
+            throws DotDataException, DotSecurityException {
+        if (userDoesNotHaveAccess(user)) {
+            throw new DotSecurityException(String.format(
+                    "Invalid service delete attempt on `%s` for host `%s` performed by user with id `%s`",
+                    serviceKey, host.getIdentifier(), user.getUserId()));
+        } else {
+            secretsStore.deleteValue(internalKey(serviceKey, host));
+        }
+    }
+
+    private static final String DESCRIPTORS_CACHE_GROUP = "DESCRIPTORS_CACHE_GROUP";
+    private static final String DESCRIPTORS_LIST_KEY = "DESCRIPTORS_LIST_KEY";
+    private static final String DESCRIPTORS_MAPPED_BY_SERVICE_KEY = "DESCRIPTORS_MAPPED_BY_SERVICE_KEY";
+
+    @Override
+    public List<ServiceDescriptor> getServiceDescriptors(User user)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Parameter 'user' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4623512)